src/ray/core_worker/actor_manager.cc

```diff
@@ -201,67 +200,80 @@ std::vector<ObjectID> ActorManager::GetActorHandleIDsFromHandles() {
-          WorkerID::FromBinary(actor_handle->GetOwnerAddress().worker_id()),
-          [this, actor_id, node_id](Status status,
-                                    const boost::optional<rpc::WorkerTableData> &result) {
-            if (!status.ok()) {
-              return;
-            }
+void ActorManager::MarkPendingActorLocationResolved(const ActorID &actor_id) {
+  absl::MutexLock lock(&mutex_);
+  auto it = actors_pending_location_resolution_.find(actor_id);
+  if (it == actors_pending_location_resolution_.end()) {
+    return;
+  }
+  actors_pending_location_resolution_.erase(it);
+}
 
-            bool worker_or_node_failed = false;
-            if (result && !result->is_alive()) {
-              worker_or_node_failed = true;
-            } else {
-              // Check node failure. We should do this because worker failure event is not
-              // reported when nodes fail.
-              const auto &optional_node_info = gcs_client_->Nodes().Get(node_id);
-              RAY_CHECK(optional_node_info)
-                  << "Node information for an actor_id, " << actor_id << " is not found.";
-              if (optional_node_info->state() ==
-                  rpc::GcsNodeInfo_GcsNodeState::GcsNodeInfo_GcsNodeState_DEAD) {
-                worker_or_node_failed = true;
-              }
-            }
+bool ActorManager::IsActorLocationPending(const ActorID &actor_id) {
+  absl::MutexLock lock(&mutex_);
+  auto it = actors_pending_location_resolution_.find(actor_id);
+  return it != actors_pending_location_resolution_.end();
+}
 
-            if (worker_or_node_failed) {
-              // We should make sure one more time that an actor is not registered to GCS
-              // before resolving actor's location to avoid race condition.
-              RAY_CHECK_OK(gcs_client_->Actors().AsyncGet(
-                  actor_id,
-                  [this, actor_id](Status status,
-                                   const boost::optional<gcs::ActorTableData> &result) {
-                    absl::MutexLock lock(&mutex_);
-                    const std::unique_ptr<ActorHandle> &actor_handle =
-                        GetActorHandleInternal(actor_id);
-                    if (status.ok() && !result && !actor_handle->IsPersistedToGCS()) {
-                      direct_actor_submitter_->DisconnectActor(actor_id, /*dead*/ true);
-                      actors_pending_location_resolution_.erase(actor_id);
-                    }
-                  }));
-            }
-          }));
-    }
-  }
+void ActorManager::DisconnectPendingLocationActorIfNeeded(const ActorID &actor_id) {
+  const std::unique_ptr<ActorHandle> &actor_handle = GetActorHandleInternal(actor_id);
+  const ClientID &node_id =
+      ClientID::FromBinary(actor_handle->GetOwnerAddress().raylet_id());
+  const WorkerID &worker_id =
+      WorkerID::FromBinary(actor_handle->GetOwnerAddress().worker_id());
+
+  RAY_CHECK_OK(gcs_client_->Workers().AsyncGet(
```

Comment:
Can you add a comment here to explain why we need to check the worker and node tables?